src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs

```diff
@@ -84,51 +103,84 @@ public Task<RestoreResult> ExecuteAsync()
+
+                        noOpTelemetry.EndIntervalMeasure(NoOpEvaluateCacheFileDuration);
+
+                        cacheFile = cacheFileAndStatus.Key;
+                        if (cacheFileAndStatus.Value)
                         {
-                            // Replay Warnings and Errors from an existing lock file in case of a no-op.
-                            await MSBuildRestoreUtility.ReplayWarningsAndErrorsAsync(_request.ExistingLockFile, _logger);
-
-                            restoreTime.Stop();
-
-                            return new NoOpRestoreResult(
-                                _success,
-                                _request.ExistingLockFile,
-                                _request.ExistingLockFile,
-                                _request.ExistingLockFile.Path,
-                                cacheFile,
-                                _request.Project.RestoreMetadata.CacheFilePath,
-                                _request.ProjectStyle,
-                                restoreTime.Elapsed);
+                            noOpTelemetry.StartIntervalMeasure();
+
+                            var noOpSuccess = NoOpRestoreUtilities.VerifyAssetsAndMSBuildFilesAndPackagesArePresent(_request);
+
+                            noOpTelemetry.EndIntervalMeasure(NoOpSuccessEvaluationDuration);
+                            noOpTelemetry.TelemetryEvent[NoOpSuccessEvaluation] = noOpSuccess;
+
+                            if (noOpSuccess)
+                            {
+                                noOpTelemetry.StartIntervalMeasure();
+
+                                // Replay Warnings and Errors from an existing lock file in case of a no-op.
+                                await MSBuildRestoreUtility.ReplayWarningsAndErrorsAsync(_request.ExistingLockFile, _logger);
+
+                                noOpTelemetry.EndIntervalMeasure(NoOpReplayLogsDuration);
+
+                                restoreTime.Stop();
+
+                                return new NoOpRestoreResult(
+                                    _success,
+                                    _request.ExistingLockFile,
+                                    _request.ExistingLockFile,
+                                    _request.ExistingLockFile.Path,
+                                    cacheFile,
+                                    _request.Project.RestoreMetadata.CacheFilePath,
+                                    _request.ProjectStyle,
+                                    restoreTime.Elapsed);
+                            }
                         }
                     }
                 }
 
-                // Restore
-                var graphs = await ExecuteRestoreAsync(
+                IEnumerable<RestoreTargetGraph> graphs = null;
+                using (TelemetryActivity.CreateTelemetryActivityWithNewOperationIdAndEvent(parentId: _operationId, eventName: ExecuteRestoreAsyncInformation))
```

Comment:
~80% of the job of restore is done here where it gather all dependencies, P2P references, resolves all dependencies, generate restore graph, and runtime graphs across TFMs. so is it possible to further segregate this? This was the complex piece which is why it wasn't done until now.

May be it'll help to come up with 4-5 separate logical pieces to cover overall restore cost and then raise these sub events?